### ISSUE TITLE
Rework Tester Statuses in TestHarness project

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#8614]
### Description of the enhancement or error report

The Status are currently a list of "defines" but there are also several attributes that go along with each status that are more or less fixed. These are scattered into the Tester, but could be encapsulated.

More importantly there's a problem with the implementation of the actual `status` in the tester itself. Each tester should have exactly one status. However, it's currently implemented as a dictionary (`{reason, status}`) which means each tester could have any number of statuses at the same time indexed by different reasons. This is definitely wrong and doesn't make any sense!

### Rationale for the enhancement or information for reproducing the error

Design of status needs to be fixed.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Internal state of the Testers, there shouldn't be any change to the API and usage in the TestHarness.
@milljm

**Comment by aeslaughter** [comment]
You can do this with https://docs.python.org/2/library/collections.html#collections.namedtuple

```
import collections 

class Tester(...):
    Status = collections.nampedtuple('Status', 'pass_message fail_message color')
    PASS = Status(pass_message='OK', fail_message='FAIL', color='RED')
    DIFF = Status(pass_massage='OK', fail_message='OK', color='YELLOW')
    PENDING = Status(pass_message='PENDING', fail_message='FAILED', color='PURPLE')

def __init__(...):

    self.status = Tester.PENDING 

def doSomethingThatCausesADiff(self):
    self.status = Tester.DIFF 


```

**Comment by permcody** [comment]
The big problem with named tuples here is that one of the items in the status is a reason string which is dynamic and changes as the test progresses. Named tuples are not mutable so setting a reason later is hackish at best. Yes it can be done if you create a new tuple from the old one replacing the status each time. A better solution is to create a small class with the 3 or 4 members in it.

**Comment by aeslaughter** [comment]
That is true, if there is dynamic content then this won't work. I was under
the impression the the status was a fixed item, where the entire status
object would change, not just the reason.

Personally, I think the reason should be baked into the status flags. If
your status is DIFF the reason is "there was a diff", if your status is
SKIP then the reason is the test was skipped. They are one in the same, but
you all can do what you want.

**Comment by permcody** [comment]
Skip is actually one of the cases where you actually do want more
information. Is it an issue, does it diff sometimes, is there a race
condition? This is the "reason" part of skip. There are several cases where
you don't want more status types but you do want an extra piece of
information that changes based on the test.

**Comment by aeslaughter** [comment]
Alright, then it seems like the reason is the only think that changes so
the class should be set to only allow that. @milljm hopefully this will
help you get started.


```
class StatusBase(object):
    def __init__(self, message, color, reason):
        self.__message = message
        self.__color = color
        self._reason = reason

   def setReason(self, reason):
        self._reason = reason

class Pass(StatusBase):
    def __init__(self, reason=None):
        super(Skip, self).__init__('OK', 'GREEN', reason)
```

Then you can just set stuff on this class:

```
status = Pass(reason='You did something right, good job!')
```

**Comment by permcody** [comment]
Closed #8684
